Sources/Workspace/DefaultPluginScriptRunner.swift

```diff
@@ -179,35 +182,80 @@ public struct DefaultPluginScriptRunner: PluginScriptRunner {
         
-            // Compile the plugin script asynchronously.
-            Process.popen(arguments: command, environment: toolchain.swiftCompilerEnvironment, queue: callbackQueue) {
-                // We are now on our caller's requested callback queue, so we just call the completion handler directly.
-                dispatchPrecondition(condition: .onQueue(callbackQueue))
-                completion($0.tryMap {
-                    // Emit the compiler output as observable info.
-                    let compilerOutput = ((try? $0.utf8Output()) ?? "") + ((try? $0.utf8stderrOutput()) ?? "")
-                    observabilityScope.emit(info: compilerOutput)
-
-                    // We return a PluginCompilationResult for both the successful and unsuccessful cases (to convey diagnostics, etc).
-                    let result = PluginCompilationResult(compilerResult: $0, diagnosticsFile: diagnosticsFile, compiledExecutable: executableFile)
-                    guard $0.exitStatus == .terminated(code: 0) else {
-                        throw DefaultPluginScriptRunnerError.compilationFailed(result)
-                    }
-                    return result
-                })
+            // Hash the command line and the contents of the source files to decide whether we need to recompile the plugin executable. We have a bounded set of input conditions, and can control what we hash.
+            let compilerInputsHash: String? = try? {
+                let stream = BufferedOutputByteStream()
+                stream <<< command
+                for sourceFile in sources.paths {
+                    try stream <<< localFileSystem.readFileContents(sourceFile).contents
+                }
+                return stream.bytes.sha256Checksum
+            }()
+
+            // If we already have a compiled executable, then compare its hash with the new one. If they're the same, we don't need to compile.
+            var compilationNeeded = true
+            let hashFile = executableFile.parentDirectory.appending(component: execName + ".inputhash")
+            if localFileSystem.exists(executableFile) {
+                if (try? localFileSystem.readFileContents(hashFile)) == compilerInputsHash {
+                    compilationNeeded = false
+                }
+            }
+            if compilationNeeded {
+                // We need to recompile the executable, so we do so asynchronously.
+                Process.popen(arguments: command, environment: toolchain.swiftCompilerEnvironment, queue: callbackQueue) {
+                    // We are now on our caller's requested callback queue, so we just call the completion handler directly.
+                    dispatchPrecondition(condition: .onQueue(callbackQueue))
+                    completion($0.tryMap {
+                        // Emit the compiler output as observable info.
+                        let compilerOutput = ((try? $0.utf8Output()) ?? "") + ((try? $0.utf8stderrOutput()) ?? "")
```

Comment:
is try? hiding error we should report?

Reply:
Not in this case.  The API here is weird; in practice this will only throw if we don't collect the output, which we do.  This could be force unwrap except that we try to avoid those.